apache/metron
# METRON-1638: Retrieve Pcap results in pdml format

## Contributor Comments
This PR adds a REST endpoint that returns pcap query results for the specified page in [PDML](https://wiki.wireshark.org/PDML) format (serialized as json instead of xml).  This will allow us to display pcap information in the Alerts UI.

A new getPdml method was added to the PcapService.  This function gets a HDFS input stream given a path that's looked up by username, job id and page number.  A Java Process object calls out to a `pcap_to_pdml.sh` script that wraps a simple tshark command.  The HDFS inputs stream is connected to standard in of the script and PDML results are read from the process output stream.  Jackson is then used to convert the output xml to json and returned.

### Changes Included

- A new PDML endpoint that accepts a job id and page number and returns PDML json for that page
- A new PcapService getPath method that returns an HDFS path given a username, job id and page number
- A new PcapService method described above
- Annotated Jackson Javabean classes for transforming PDML xml to json
- A `pcap_to_pdml.sh` script that wraps a simple tshark command and can be extended as needed
- A new script path Spring property that can be changed to point to a different script location
- Unit and integration tests

### Testing

This has been tested in full dev.  Testing instructions are as follows:

1. Spin up full dev
2. Create the HDFS directories specified in application.yml
  - /apps/metron/pcap/input
  - /apps/metron/pcap/interim
  - /apps/metron/pcap/output
3. Put pcap data in `/apps/metron/pcap/input`
4. Run a fixed pcap query to generate results:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "endTimeMs": 1458240269424,
  "startTimeMs": 1458240269419
}' 'http://node1:8082/api/v1/pcap/fixed'
```
5. Monitor the job with the get status endpoint until the job has SUCCEEDED and returns a pageTotal:
```
{
  "jobId": "job_1532012905403_0001",
  "jobStatus": "SUCCEEDED",
  "description": "map: 20.0%, reduce: 0.0%",
  "percentComplete": 100,
  "pageTotal": 2
}
```
6. Attempt to get PDML results using the job id and page number:
```
curl -X GET --header 'Accept: application/json' 'http://node1:8082/api/v1/pcap/job_1532012905403_0001/pdml?page=1'
```
You should get an error message that clearly states wireshark is missing:
```
{
  "responseCode": 500,
  "message": "/usr/metron/0.5.1/bin/pcap_to_pdml.sh: line 19: tshark: command not found\n",
  "fullMessage": "RestException: /usr/metron/0.5.1/bin/pcap_to_pdml.sh: line 19: tshark: command not found\n"
}
```
7.  Install wireshark in full dev with `yum -y install wireshark`.  This time PDML should be returned:
```
{
  "version": "0",
  "creator": "wireshark/1.8.10",
  "time": "Thu Jul 19 16:47:41 2018",
  "captureFile": "",
  "packets": [
    {
      "protos": [
        {
          "name": "geninfo",
          "pos": "0",
          "showname": "General information",
          "size": "130",
          "hide": null,
          "fields": [
            {
              "name": "num",
...
}
```
8.  Attempting to get PDML results for pages that don't exist (outside of the pageTotal) should return a 404.

I'm still working on adding some documentation to the script and endpoint.  Otherwise this PR is ready for review.  Expect another doc commit shortly.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
